exporters/prometheus/exporter_test.go

```diff
@@ -0,0 +1,110 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package prometheus
+
+import (
+	"context"
+	"net/http/httptest"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+)
+
+func TestPrometheusExporter(t *testing.T) {
+	ctx := context.Background()
+	labels := []attribute.KeyValue{
+		attribute.Key("A").String("B"),
+		attribute.Key("C").String("D"),
+	}
+
+	reader := metric.NewManualReader()
+	provider := metric.NewMeterProvider(metric.WithReader(reader))
+	exporter, err := NewPrometheusExporter(reader)
+	require.NoError(t, err)
+
+	expectedLines := []string{""}
+
+	meter := provider.Meter("testmeter")
+	counter, err := meter.SyncFloat64().Counter("foo")
+	require.NoError(t, err)
+	counter.Add(ctx, 5, labels...)
+	counter.Add(ctx, 10.3, labels...)
+	counter.Add(ctx, 9, attribute.Key("X").String("Y"))
+	expectedLines = append(expectedLines,
+		`# HELP foo `,
+		`# TYPE foo counter`,
+		`foo{X="Y"} 9`,
+		`foo{A="B",C="D"} 15.3`,
+	)
+
+	gauge, err := meter.SyncFloat64().UpDownCounter("bar")
+	require.NoError(t, err)
+	gauge.Add(ctx, 100, labels...)
+	gauge.Add(ctx, -25, labels...)
+	expectedLines = append(expectedLines,
+		`# HELP bar `,
+		`# TYPE bar counter`,
+		`bar{A="B",C="D"} 75`,
+	)
+
+	histogram, err := meter.SyncFloat64().Histogram("baz")
+	require.NoError(t, err)
+	histogram.Record(ctx, 23, labels...)
+	histogram.Record(ctx, 7, labels...)
+	histogram.Record(ctx, 101, labels...)
+	histogram.Record(ctx, 105, labels...)
+	expectedLines = append(expectedLines,
+		`# HELP baz `,
+		`# TYPE baz histogram`,
+		`baz_bucket{A="B",C="D",le="0"} 0`,
+		`baz_bucket{A="B",C="D",le="5"} 0`,
+		`baz_bucket{A="B",C="D",le="10"} 1`,
+		`baz_bucket{A="B",C="D",le="25"} 1`,
+		`baz_bucket{A="B",C="D",le="50"} 0`,
+		`baz_bucket{A="B",C="D",le="75"} 0`,
+		`baz_bucket{A="B",C="D",le="100"} 0`,
+		`baz_bucket{A="B",C="D",le="250"} 2`,
+		`baz_bucket{A="B",C="D",le="500"} 0`,
+		`baz_bucket{A="B",C="D",le="1000"} 0`,
+		`baz_bucket{A="B",C="D",le="+Inf"} 4`,
+		`baz_sum{A="B",C="D"} 236`,
+		`baz_count{A="B",C="D"} 4`,
+	)
+
+	rec := httptest.NewRecorder()
+	req := httptest.NewRequest("GET", "/metrics", nil)
+	promhttp.HandlerFor(exporter.registry, promhttp.HandlerOpts{}).ServeHTTP(rec, req)
```

Comment:
I'd recommend using [testutil.GatherAndCompare](https://github.com/prometheus/client_golang/blob/1638da9ae43be6ff5ad522ca402712e26b4a19c7/prometheus/testutil/testutil.go#L201), and keeping the expected output for each case in a text file.

Reply:
I tried doing this in my latest commit (pushed separate) but the compare function is failing in a weird way. Maybe I'm doing something wrong opening the files:
```
--- FAIL: TestPrometheusExporter (0.00s)
    --- FAIL: TestPrometheusExporter/counter (0.00s)
        exporter_test.go:108: 
            	Error Trace:	exporter_test.go:108
            	Error:      	Received unexpected error:
            	            	
            	            	
            	            	Diff:
            	            	--- metric output does not match expectation; want
            	            	+++ got:
            	            	@@ -1,2 +1,3 @@
            	            	-(bytes.Buffer) # TYPE foo counter
            	            	+(bytes.Buffer) # HELP foo 
            	            	+# TYPE foo counter
            	            	 foo{A="B",C="D"} 24.3
            	Test:       	TestPrometheusExporter/counter
    --- FAIL: TestPrometheusExporter/gauge (0.00s)
        exporter_test.go:108: 
            	Error Trace:	exporter_test.go:108
            	Error:      	Received unexpected error:
            	            	
            	            	
            	            	Diff:
            	            	--- metric output does not match expectation; want
            	            	+++ got:
            	            	@@ -1,2 +1,3 @@
            	            	-(bytes.Buffer) # TYPE bar counter
            	            	+(bytes.Buffer) # HELP bar 
            	            	+# TYPE bar counter
            	            	 bar{A="B",C="D"} 75
            	Test:       	TestPrometheusExporter/gauge
    --- FAIL: TestPrometheusExporter/histogram (0.00s)
        exporter_test.go:108: 
            	Error Trace:	exporter_test.go:108
            	Error:      	Received unexpected error:
            	            	
            	            	
            	            	Diff:
            	            	--- metric output does not match expectation; want
            	            	+++ got:
            	            	@@ -1,2 +1,3 @@
            	            	-(bytes.Buffer) # TYPE baz histogram
            	            	+(bytes.Buffer) # HELP baz 
            	            	+# TYPE baz histogram
            	            	 baz_bucket{A="B",C="D",le="0"} 0
            	Test:       	TestPrometheusExporter/histogram
FAIL
FAIL	go.opentelemetry.io/otel/exporters/prometheus	0.072s
FAIL
make: *** [Makefile:115: test/./exporters/prometheus] Error 123
```

Reply:
What's really weird is
```
            	Error:      	Received unexpected error:
            	            	
            	            	
            	            	Diff:
            	            	--- metric output does not match expectation; want
```
where it looks like the error message is being printed in the middle of the diff output. maybe something weird with buffering

Reply:
Maybe its a white-space diff?  Try adding a space after `# HELP foo`, for example.  Or just add a description in the test.

Here is an example i've worked on https://github.com/google/cadvisor/blob/master/metrics/prometheus_test.go#L78.    Example metrics test file: https://github.com/google/cadvisor/blob/master/metrics/testdata/prometheus_machine_metrics

Reply:
Must be a whitespace thing, adding a description worked (I had already tried adding an extra space, which didn't work, but maybe `os.Open()` trims extra whitespace or something)